File: app/src/main/java/com/github/gfx/helium/widget/ArrayRecyclerAdapter.java
package com.github.gfx.helium.widget;

import android.content.Context;
import android.support.annotation.NonNull;
import android.support.annotation.UiThread;
import android.support.v7.widget.RecyclerView;
import android.view.View;

import java.util.ArrayList;
import java.util.Collection;
import java.util.Iterator;

public abstract class ArrayRecyclerAdapter<T, VH extends RecyclerView.ViewHolder>
        extends RecyclerView.Adapter<VH> implements Iterable<T> {

    final Context context;

    final ArrayList<T> list;

    OnItemClickListener<T> itemClickListener;

    OnItemLongClickListener<T> itemLongClickListener;

    public ArrayRecyclerAdapter(@NonNull Context context) {
        this.context = context;
        this.list = new ArrayList<>();
    }

    @Override
    public int getItemCount() {
        return list.size();
    }

    @UiThread
    public void reset(Collection<T> items) {
        clear();
        addAll(items);
        notifyDataSetChanged();
    }

    public T getItem(int position) {
        return list.get(position);
    }

    public void addItem(T item) {
        list.add(item);
    }

    public void addAll(Collection<T> items) {
        list.addAll(items);
    }

    public void addAll(int position, Collection<T> items) {
        list.addAll(position, items);
    }

    @UiThread
    public void addAllWithNotification(Collection<T> items) {
        int position = getItemCount();
        addAll(items);
        notifyItemInserted(position);
    }

    public void clear() {
        list.clear();
    }

    public Context getContext() {
        return context;
    }

    public void setOnItemClickListener(OnItemClickListener<T> listener) {
        itemClickListener = listener;
    }

    public void setOnItemLongClickListener(OnItemLongClickListener<T> listener) {
        itemLongClickListener = listener;
    }

    public void dispatchOnItemClick(View view, T item) {
        assert itemClickListener != null;
        itemClickListener.onItemClick(view, item);
    }

    public boolean dispatchOnItemLongClick(View view, T item) {
        assert itemLongClickListener != null;
        return itemLongClickListener.onItemLongClick(view, item);
    }

    @Override
    public Iterator<T> iterator() {
        return list.iterator();
    }
}

File: app/src/main/java/com/github/gfx/helium/widget/ArrayRecyclerAdapter.kt
package com.github.gfx.helium.widget

import android.content.Context
import android.support.annotation.UiThread
import android.support.v7.widget.RecyclerView
import android.view.View

import java.util.ArrayList

abstract class ArrayRecyclerAdapter<T, VH : RecyclerView.ViewHolder>(val context: Context) : RecyclerView.Adapter<VH>(), Iterable<T> {

    internal val list: ArrayList<T>

    internal var itemClickListener: OnItemClickListener<T>? = null

    internal var itemLongClickListener: OnItemLongClickListener<T>? = null

    init {
        this.list = ArrayList<T>()
    }

    override fun getItemCount(): Int {
        return list.size
    }

    @UiThread
    fun reset(items: Collection<T>) {
        clear()
        addAll(items)
        notifyDataSetChanged()
    }

    fun getItem(position: Int): T {
        return list[position]
    }

    fun addItem(item: T) {
        list.add(item)
    }

    fun addAll(items: Collection<T>) {
        list.addAll(items)
    }

    fun addAll(position: Int, items: Collection<T>) {
        list.addAll(position, items)
    }

    @UiThread
    fun addAllWithNotification(items: Collection<T>) {
        val position = itemCount
        addAll(items)
        notifyItemInserted(position)
    }

    fun clear() {
        list.clear()
    }

    fun setOnItemClickListener(listener: OnItemClickListener<T>) {
        itemClickListener = listener
    }

    fun setOnItemLongClickListener(listener: OnItemLongClickListener<T>) {
        itemLongClickListener = listener
    }

    fun dispatchOnItemClick(view: View, item: T) {
        assert(itemClickListener != null)
        itemClickListener!!.onItemClick(view, item)
    }

    fun dispatchOnItemLongClick(view: View, item: T): Boolean {
        assert(itemLongClickListener != null)
        return itemLongClickListener!!.onItemLongClick(view, item)
    }

    override fun iterator(): Iterator<T> {
        return list.iterator()
    }
}
